### DISCUSSION TITLE
I am making an ImageMagick wrapper in PowerShell for my own purposes. I need help coming up with a way to determine the correct density to use for any given file.

### DISCUSSION THREAD

**Discussion by futuremotiondev** [ImageMagick/ImageMagick#6197] [Help]
### ImageMagick version

7.1.0-48 Q16-HDRI x64

### Operating system, version and so on

Windows 10 Pro x64 22H2

### Description

I have been testing the conversion of various vector images to raster formats as I am in the process of writing a simple wrapper around IM for my own personal use. The application is not at all aimed for print use, only screen.

The application is simple: You load a image, type in dimensions in pixels, select your destination format, and finally press convert and let it rip. 

This is the general format I'm using for AI and PDF files (I have a lot of variations for different operations):

`magick convert -colorspace sRGB -density 600 -depth 8 $aiPath1 -background none -resize 1000x1000 -units pixelsperinch $pngPath`

The issue I'm running into is that the initial density needs to be scaled by "some" factor to ensure crisp rendering when the results are saved. 

For instance, I have a small file called `Shields.ai` (Identify reports this file's geometry as 40x40). In order to get a crisp result for this image at an output size of `1000px`, the initial `-density` value needs to be cranked up to about `2200`. 

Another file that's `512x512` only needs an initial density of `600` to get a crisp render at `2000px`.

Yet another file with several `12x12px` artboards needs a density of `5000` to get a crisp render at 1000px. 

So the reason I'm posting here is that I need help determining some general rules that I can translate to code so that a proper density value is used for the current file. I also don't quite understand how `Density`, `Geometry`, `Resolution`, and `Print size` all interact. From my initial testing it seems like that geometry size is the most important property to take into consideration when specifying the density of the image magick call.

Initially I thought to create "tiers" of density values based on reported geometry and assign the file to the correct tier (density value) but I don't feel like I have a proper grasp of the fundamentals here.

I also thought about trying to find some kind of optimal scale factor using the reported geometry of the image being converted.

I'm hoping someone can steer me in a better direction.

Thanks for reading, sorry for the length.

TL;DR: What is an optimal way to compute density when converting vector sources to raster at arbitrary resolutions?

**Comment by snibgo** [comment]
For your purposes, use "`-density`" before reading the AI or PDF file.

As a rule of thumb, AI and PDF files contain vector data, and are about 8 by 11 inches, so a density of about 72 DPI works fairly well, resulting in raster images 576x792 pixels. Resizing these up to 1000x1000 won't look good. You should generally specify a density that gives more pixels then you need (eg 144 dpi), then resize down.

IM won't tell you the size of the AI file without rasterizing it. So a common method is to rasterize twice. The first time at some known dpi such as 72. A calculation then tells you what density will give you the number of pixels you want.

Sadly, some input vector files are weird, and rasterizing these at 72 dpi can result in far too many pixels, perhaps too many for your computer to handle easily. To cope with these files, you may want to first rasterize at a very low density such as 1 dpi.

You might experiment with "`-define pdf:fit-page=1000x1000`" before reading the AI. This seems to set an upper limit on the size of the raster image.

**Comment by futuremotiondev** [comment]
Thanks so much for the help.

> As a rule of thumb, AI and PDF files contain vector data, and are about 8 by 11 inches, so a density of about 72 DPI works fairly well, resulting in raster images 576x792 pixels. Resizing these up to 1000x1000 won't look good.

I'm familiar with AI and PDF formats. I do a lot of icon design and have a lot of vector files that aren't your typical "A4/A5/A6/whatever document". Often I want to blow up an icon or design to some arbitrary large size. In my application requirements, I want the end user to be able to specify dimensions for the longest edge so I can bake down different PNG sizes in batch.

> You should generally specify a density that gives more pixels then you need (eg 144 dpi), then resize down.

Yes, absolutely. That's exactly what I'm trying to do. My exact question and purpose of this post is how to arrive at the optimal density value that "gives more pixels then I need". The key word is _need_. How do I know what I need? It's easy if you're working with standard page sizes all the time, but in my case it's a bit more complex.

After running experiments on several different AI/EPS files it seems like the geometry value supplied by identify is the biggest factor.

> IM won't tell you the size of the AI file without rasterizing it. So a common method is to rasterize twice. The first time at some known dpi such as 72. A calculation then tells you what density will give you the number of pixels you want.

This is trivial for me to solve in PowerShell, I can just call identify first and then extract the geometry size > store into a variable > compute optimal density.

I ran some experiments with a `-resize` value of `1000x1000px` and came to these values. The left is actual geometry size in pixels, the right is the density value that I considered "optimal" after comparing.

```
12x12     6000
40x40     2500
60x60     1800-1900
100x100   850
200x200   500-600
250x250   400-500
512x512   400-500
600x337   400-500
720x1104  400-500
800x600   400-500
1500x1500 300
2500x1500 200-300
```
Since this app will be deployed locally on my machine for general use, I want it to be able to handle any resize operation gracefully and always produce **crisp** output.

> Sadly, some input vector files are weird, and rasterizing these at 72 dpi can result in far too many pixels, perhaps too many for your computer to handle easily. To cope with these files, you may want to first rasterize at a very low density such as 1 dpi.

Yeah, I'm expecting some hiccups like that along the way. I wonder if a call to identify first could detect these edge cases somehow. If you have any "messed up" vector files that don't behave properly please send them my way! I'd like to experiment to see if I can get some proper error handling set up. 

Thanks again for your input.

**Comment by snibgo** [comment]
> ... that "gives more pixels then I need". The key word is *need*. How do I know what I need?

I can't answer that for you. But once you have an answer, the processing is then fairly simple. Suppose you want at least 1000 pixels on each side. Then you first rasterize at 72 dpi. Suppose the result is 400x500 pixels. The smaller dimension is 400. You need to re-rasterize at 72 * 1000/400 dpi, 180 dpi.

> If you have any "messed up" vector files that don't behave properly please send them my way!

Cases are often raised on these forums, for example the recent post https://github.com/ImageMagick/ImageMagick/discussions/6184.
